portfolio/src/main/java/com/google/sps/servlets/MarkersServlet.java

```diff
@@ -0,0 +1,115 @@
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.gson.Gson;
+import com.google.sps.data.DataUtil;
+import com.google.sps.data.MarkerData;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.jsoup.Jsoup;
+import org.jsoup.safety.Whitelist;
+
+/** Handles fetching and saving markers data. */
+@WebServlet("/markers")
+public class MarkersServlet extends HttpServlet {
+
+  /** Responds with a JSON array containing marker data. */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+
+    Collection<MarkerData> markersData = getMarkersData();
+    Gson gson = new Gson();
+    String json = gson.toJson(markersData);
+
+    response.getWriter().println(json);
+  }
+
+  /** Accepts a POST request containing a new marker. */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    // User can only comment if they are logged in
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/index.html");
+      return;
+    }
+
+    String id = userService.getCurrentUser().getUserId();
+    String displayName = request.getParameter("name");
+    
+    // Update user's displayName in DataStore
+    Entity userInfoEntity = new Entity("UserInfo", id);
+    userInfoEntity.setProperty("id", id);
+    userInfoEntity.setProperty("displayName", displayName); 
+    
+    double lat = Double.parseDouble(request.getParameter("lat"));
+    double lng = Double.parseDouble(request.getParameter("lng"));   
+    String content = Jsoup.clean(request.getParameter("content"), Whitelist.none());
+
+    // Place information about marker in Datastore
+    Entity markerEntity = new Entity("MarkerDataEntity");
+    markerEntity.setProperty("lat", lat);
+    markerEntity.setProperty("lng", lng);
+    markerEntity.setProperty("id", id);
+    markerEntity.setProperty("content", content);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(markerEntity);
+    datastore.put(userInfoEntity);
+  }
+
+  /** Fetches markers from Datastore. */
+  private Collection<MarkerData> getMarkersData() {
+    Collection<MarkerData> markersData = new ArrayList<>();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("MarkerDataEntity");
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      double lat = (double) entity.getProperty("lat");
```

Comment:
I feel like for here and above, it would be nice to have conversion functions.

ex. MarkerData createMarkerData(Entity entity);

Or maybe just have another constructor for MarkerData that takes in an Entity?

WDYT?

Reply:
I think that's a good idea! I have added this in the latest commit.